File: Chromium/silver/precomp.rb
require 'set'
require 'matrix'

class Array
  def bs_idx
    low = 0; high = length - 1
    until low > high
      mid = low + ((high - low) / 2)
      cmp = yield(self[mid])
      if cmp == 1
        high = mid - 1
      elsif cmp == -1
        low = mid + 1
      elsif cmp == 0
        return [true, mid]
      end
    end
    return [false, low]
  end
end


def count_stack(switches, height=switches.first, max=switches.last)
  return (max + 1 - height) unless switches.any?
  switches = [height] + switches
  left  = 1
  right = 1
  total = 1
  path = 0
  last_height = height

  # branch = $tree
  # switches.each do |switch|
  #   branch = branch[switch - switches[0]]
  # end
  switches.each do |level|
    levels = level - last_height
    next if levels.zero?
    if path == 1
      increase = left * levels
      path = 0
      total = (total + (increase)) % 1_000_000_007
      right = (right + increase)% 1_000_000_007
    else
      increase = right * levels
      path = 1
      total = (total +(increase)) % 1_000_000_007
      left  = (left +increase) % 1_000_000_007
    end
    last_height = level
  end
  levels = max - switches[-1]
  total += ((path == 1 ? left : right) * levels) % 1_000_000_007
  return total
end

def solution2(values)
  puts "Start"
  indices_by_height = values.each.with_index.sort_by(&:first).map.with_index(1) do |(height, position), index|
    [index, position]
  end.to_h

  heights_by_index = indices_by_height.invert
  puts "Initial sort"
  switches = Array.new(values.size)
  removes = Array.new(values.size)

  position = indices_by_height[1]

  indices_by_height.to_a[1..-1].each do |height, next_position|
    direction = next_position < position ? :left : :right
    case direction
    when :left
      # binding.pry
      switches[next_position] ||= Set.new
      switches[next_position] << height - 1
      removes[position] ||= Set.new
      removes[position] << height - 1
    when :right
      switches[position] ||= Set.new
      switches[position] << height - 1
      removes[next_position] ||= Set.new
      removes[next_position] << height - 1
    end
    position = next_position
  end

  puts "Build switches"
  current_switches = []

  switches.each.with_index.reduce(0) do |total, (node_switches, position)|
    print "\r#{position}"
    height = heights_by_index[position]

    to_remove = removes[position]
    change_index = nil
    (to_remove || []).each do |remove|
      found, remove_index = current_switches.bs_idx{|x| x <=> remove}
      if found
        current_switches.delete_at(remove_index)
      end
    end

    if node_switches
      node_switches.each do |switch|
        _, insert_index = current_switches.bs_idx{|x| x <=> switch }
        current_switches.insert(insert_index, switch)
      end
    end

    found, insert_index = current_switches.bs_idx{|x| x <=> height }
    applicable = current_switches[(found ? (insert_index + 1): insert_index)..-1]

    # matrix = (applicable + [values.length]).each_cons(4).map{|l1, r1, l2, r2| get_matrix([r1 - l1, r2 - l1])}.inject(:*)
    # matrix = get_matrix(pairs)
    # binding.pry
    # puts "S#{height}: #{count_stack(height, applicable, values.length)}"
    step = count_stack(applicable, height, values.length) % 1_000_000_007
    # puts "Step #{step}"
    # print "\r#{position}: #{step}"
    (total + step) % 1_000_000_007
  end
end

def find_unequal
  (3..9).each do |i|
    values = i.times.with_index(1).to_a.map(&:last)
    values.permutation.each do |values|
      return values if solution2(values) != solution(values)
    end
  end
end
# values = [2,1,3,4]
# puts solution(values)
# puts solution2(values)
# puts "Unequal #{find_unequal}"
# values = [4,1,3,6]
#

values = 50_000.times.map{|i| i}.shuffle
# starts = Time.now
# puts solution2(values)
# ends = Time.now
# puts "Took #{ends - starts}"
starts = Time.now
puts solution2(values)
ends = Time.now
puts "Took #{ends - starts}"

File: Chromium/precomp.go
package main

import "fmt"
import "sort"
import "math/rand"

type Level struct {
    index int
    height int
}

type Levels []Level


func (l Level) String() string {
  return fmt.Sprintf("(%d %d)", l.index, l.height)
}

func (s Levels) Len() int {
    return len(s)
}

func (s Levels) Swap(i, j int) {
    s[i], s[j] = s[j], s[i]
}
func (s Levels) Less(i, j int) bool {
    return s[i].height < s[j].height
}

func WithIndices(vs []int) []Level {
    vsm := make([]Level, len(vs))
    for i, v := range vs {
        vsm[i] = Level{i, v}
    }
    return vsm
}

func Insert(slice []int, index, value int) []int {
    slice = slice[0 : len(slice)+1]
    copy(slice[index+1:], slice[index:])
    slice[index] = value
    return slice
}

func Remove(slice []int, index int) []int{
  length := len(slice)
  slice = append(slice[:index], slice[index+1:]...)
  return slice[:length - 1]
}

func FindIndex(target_map []int, value int) (int, bool) {

  start_index := 0
  end_index := len(target_map) - 1

  for start_index <= end_index {

    median := (start_index + end_index) / 2

    if target_map[median] < value {
      start_index = median + 1
    } else {
      end_index = median - 1
    }
  }

  if start_index == len(target_map) || target_map[start_index] != value {
    return start_index, false
  } else {
    return start_index, true
  }
}

func countStack(switches []int,height int, max int) int{
  if len(switches) == 0 {
    return max + 1 - height
  }
  left  := 1
  right := 1
  total := 1
  path  := 0
  lastHeight := height
  levels := 0
  for _, level := range switches {
    levels = level - lastHeight
    if levels == 0 {
      continue
    }
    if path == 1 {
      increase := left * levels
      path = 0
      total = (total + increase) % 1000000007
      right = (right + increase) % 1000000007
    } else {
      increase := right * levels
      path = 1
      total = (increase + total) % 1000000007
      left =  (increase + left) % 1000000007
    }
    lastHeight = level
  }
  levels = max - switches[len(switches) - 1]
  if path == 1 {
    total += (left * levels) % 1000000007
  }else {
    total += (right * levels) % 1000000007
  }
  return total
}

func Solution(A []int) int{
  preWithPIndices := WithIndices(A)
  byHeight := make(map[int]Level)
  byIndex  := make(map[int]Level)
  switches := make(map[int][]int)
  removes  := make(map[int][]int)

  sort.Sort(Levels(preWithPIndices))


  withIndices := make([]Level, len(preWithPIndices))
  for i, v := range preWithPIndices {
    withIndices[i] = Level{v.index, i + 1}
  }

  index := withIndices[0].index
  byIndex[index] = withIndices[0]

  for _, l := range withIndices[1:] {
    byHeight[l.height] = l
    byIndex[l.index] = l
    if l.index < index {
      switches[l.index] = append(switches[l.index], l.height - 1)
      removes[index]    = append(removes[index], l.height - 1)
    } else {
      switches[index]  = append(switches[index], l.height - 1)
      removes[l.index] = append(removes[l.index], l.height - 1)
    }
    index = l.index
  }

  currentSwitches  := make([]int, 50000)
  currentSwitches = currentSwitches[0:0]

  total := 0

  for i, _ := range withIndices {
    height := byIndex[i].height
    if removes, ok := removes[i]; ok{
      for _, remv := range removes {
        if index, found := FindIndex(currentSwitches, remv); found {
          currentSwitches = Remove(currentSwitches, index)
        }
      }
    }

    if switches, ok := switches[i]; ok {
      for _, swtch := range switches {
        index, _ := FindIndex(currentSwitches, swtch)
        currentSwitches = Insert(currentSwitches, index, swtch)
      }
    }

    insertIndex, found := FindIndex(currentSwitches, height)
    if found {
      insertIndex += 1
    }
    applicable := currentSwitches[insertIndex:]
    step := countStack(applicable, height, len(A))
    total = (total + step) % 1000000007
  }
  return total % 1000000007
}

func main() {
  // values := []int{1,11,5,8,7,10,3,9,6,4,2,44,36,27,81,95,41}
  values := rand.Perm(50000)
  // fmt.Printf("Values", values)
  fmt.Printf("%d", Solution(values))
}